optuna/storages/grpc/api.proto

```diff
@@ -0,0 +1,436 @@
+    int64 trial_id = 1;
+}
+
+/**
+ * Request to set a trial parameter.
+ */
+message SetTrialParameterRequest {
+    int64 trial_id = 1;
+    string param_name = 2;
+    double param_value_internal = 3;
+    string distribution = 4;
+}
+
+/**
+ * Reply to set a trial parameter.
+ */
+message SetTrialParameterReply {}
+
+/**
+ * Request to get a trial id from its study id and trial number.
+ */
+message GetTrialIdFromStudyIdTrialNumberRequest {
+    int64 study_id = 1;
+    int64 trial_number = 2;
+}
+
+/**
+ * Reply to get a trial id from its study id and trial number.
+ */
+message GetTrialIdFromStudyIdTrialNumberReply {
+    int64 trial_id = 1;
+}
+
+/**
+ * Request to set trial state and values.
+ */
+message SetTrialStateValuesRequest {
+    int64 trial_id = 1;
+    TrialState state = 2;
+    repeated double values = 3;
+}
+
+/**
+ * Reply to set trial state and values.
+ */
+message SetTrialStateValuesReply {
+    bool trial_updated = 1;
+}
+
+/**
+ * Request to set a trial intermediate value.
+ */
+message SetTrialIntermediateValueRequest {
+    int64 trial_id = 1;
+    int64 step = 2;
+    double intermediate_value = 3;
+}
+
+/**
+ * Reply to set a trial intermediate value.
+ */
+message SetTrialIntermediateValueReply {}
+
+/**
+ * Request to set a trial user attribute.
+ */
+message SetTrialUserAttributeRequest {
+    int64 trial_id = 1;
+    string key = 2;
+    string value = 3;
+}
+
+/**
+ * Reply to set a trial user attribute.
+ */
+message SetTrialUserAttributeReply {}
+
+/**
+ * Request to set a trial system attribute.
+ */
+message SetTrialSystemAttributeRequest {
+    int64 trial_id = 1;
+    string key = 2;
+    string value = 3;
+}
+
+/**
+ * Reply to set a trial system attribute.
+ */
+message SetTrialSystemAttributeReply {}
+
+/**
+ * Request to get a trial by its ID.
+ */
+message GetTrialRequest {
+    int64 trial_id = 1;
+}
+
+/**
+ * Reply to get a trial by its ID.
+ */
+message GetTrialReply {
+    FrozenTrial frozen_trial = 1;
+}
+
+/**
+ * Request to get all trials in a study.
+ */
+message GetAllTrialsRequest {
+    int64 study_id = 1;
+    repeated TrialState states = 2;
+}
+
+/**
+ * Reply to get all trials in a study.
+ */
+message GetAllTrialsReply {
+    repeated FrozenTrial frozen_trials = 1;
+}
+
+/**
+ * Study direction.
+ */
+enum StudyDirection {
+    MINIMIZE = 0;
+    MAXIMIZE = 1;
+}
+
+/**
+ * Trial state.
+ */
+enum TrialState {
+    RUNNING = 0;
+    COMPLETE = 1;
+    PRUNED = 2;
+    FAIL = 3;
+    WAITING = 4;
+}
+
+/**
+ * Frozen study.
+ */
+message FrozenStudy {
+    int64 study_id = 1;
+    string study_name = 2;
+    repeated StudyDirection directions = 3;
+    map<string, string> user_attributes = 4;
+    map<string, string> system_attributes = 5;
+}
+
+/**
+ * Frozen trial.
+ */
+message FrozenTrial {
+    int64 trial_id = 1;
+    int64 number = 2;
+    TrialState state = 3;
+    repeated double values = 4;
+    string datetime_start = 5;
+    string datetime_complete = 6;
```

Comment:
It seems common to use `google.protobuf.Timestamp` for timestamp rather than `string`.
https://github.com/search?q=google.protobuf.timestamp+language%3A%22Protocol+Buffers%22+&type=code